Add basic wlr-output-management support

joggee-fr opened the pull request:
I have tried to resume work from @dimkr in PR #213.
- Rebased on current HEAD
- Implements output manager test (`wlr-randr --dryrun`)
- Add output layout change monitor to ensure views are well positioned according to new layout

For the moment, I have only played with Wayland as backend.
I am still unsure about `current_mode` test, see [discussion](https://github.com/cage-kiosk/cage/pull/213#discussion_r1267768236).
Furthermore, trying to disable an output returns an error. I don't know if it specific to Wayland backend. Here are the logs from wlroots (0.16.2):
```
00:00:04.247 [types/output/output.c:656] Tried to modeset a disabled output
00:00:04.247 [types/output/output.c:734] Basic output test failed for WL-2
```

joggee-fr commented:
When disabling an output, we must avoid to set other parameters (scale, mode, etc.) as it will cause the error described in my first comment. Now, with a Wayland backend, we still observe an error but a kind of expected one:
`[backend/wayland/output.c:254] Unsupported output state fields: 0x8`.

I have also tried with DRM backend. Disabling output is OK but re-enabling next is not yet very stable.

joggee-fr commented:
> I have also tried with DRM backend. Disabling output is OK but re-enabling next is not yet very stable.

Simply because I forgot Cage terminates when no more output is available.

emersion commented:
> with a Wayland backend, we still observe an error but a kind of expected one

Right. With wlroots 0.16, only the DRM backend supports disabling outputs. With wlroots 0.17, all backends support disabling outputs.

emersion commented on output.c:
```diff
@@ -75,6 +100,46 @@ output_disable(struct cg_output *output)
+
+	/* Apply output configuration */
+	if (head->state.enabled) {
+		wlr_output_layout_add(output->server->output_layout, head->state.output, head->state.x, head->state.y);
```
We should add the output to the layout only after `wlr_output_commit()` succeeds. If the commit fails, we don't want to apply the rest of the configuration changes.

emersion commented on output.c:
```diff
@@ -257,3 +321,62 @@ output_set_window_title(struct cg_output *output, const char *title)
+		struct cg_output *output;
+		bool ret = false;
+
+		wl_list_for_each (output, &server->outputs, link) {
```
We don't need to iterate over the list of outputs here: we can grab the output via:

```c
struct cg_output *output = head->state.output->data;
```

emersion commented on output.c:
```diff
@@ -75,6 +100,46 @@ output_disable(struct cg_output *output)
 	wlr_output_layout_remove(output->server->output_layout, wlr_output);
 	wlr_output_commit(wlr_output);
+
+	update_output_manager_config(output->server);
```
We shouldn't need to call `update_output_manager_config()` here, the `wlr_output_commit()` call above will result in a commit signal being fired.

joggee-fr commented on output.c:
```diff
@@ -143,17 +219,11 @@ output_destroy(struct cg_output *output)
-		wl_display_terminate(server->wl_display);
-	} else if (server->output_mode == CAGE_MULTI_OUTPUT_MODE_LAST) {
+	if (server->output_mode == CAGE_MULTI_OUTPUT_MODE_LAST) {
 		struct cg_output *prev = wl_container_of(server->outputs.next, prev, link);
```
@emersion, @Hjdskes 
I have tested `-m last` option to check modifications done in `output_disable()` and have observed several segfaults.
Reading the code and checking `wl_list` API, I feel that this is not the right way to reliably retrieve the next item in list. It seems `next` could never be `NULL` and the code can consequently try to release incorrect part of the memory.

By the way, is the `-m` option still relevant if wlr-output-managent support is added?

emersion replied:
Yeah, it's not safe to use `wl_container_of()` without checking first that the pointer passed to it is `!= &server->outputs`.

joggee-fr commented:
I have added commits (to be squashed if OK) to answer @emersion comments.
By the way, I have also modified the access to previous or next item in output list causing segfaults when using `-m last` option.

emersion commented on output.c:
```diff
@@ -225,12 +225,10 @@ output_destroy(struct cg_output *output)
 	free(output);
 
-	if (server->output_mode == CAGE_MULTI_OUTPUT_MODE_LAST) {
+	if ((server->output_mode == CAGE_MULTI_OUTPUT_MODE_LAST) && !wl_list_empty(&server->outputs)) {
```
Nit: these parentheses are unnecessary

emersion commented:
Yeah, the changes you made LGTM.

emersion commented:
I'm still not super sure about "output: do not terminate when last output is destroyed". It sounds like this is useful mainly when running nested under the X11 and Wayland backends, to stop running when the last cage window is closed.

joggee-fr commented:
Squashed some commits and removed extra parentheses.

joggee-fr commented:
> I'm still not super sure about "output: do not terminate when last output is destroyed". It sounds like this is useful mainly when running nested under the X11 and Wayland backends, to stop running when the last cage window is closed.

I don't have done any tests with X11 backend for now.
When using Wayland backend, how can we destroy output? I understand with next version of wlroots, we will be able to disable outputs, but how about output destruction?

Below a short description of the issue I observed with terminate when last output available is destroyed:
- Cage started with a monitor plugged in HDMI output. DRM backend is selected, GPU is integrated Intel Iris with i915 kernel driver ;
- `HDMI-A-1` output is disabled using `wlr-randr --output HDMI-A-1 --off` ;
- Monitor lost signal and automatically enters a standby mode ;
- HDMI output is re-enabled using `wlr-randr --output HDMI-A-1 --on` ;
- I think it is due to monitor exiting "slowly" from standby but Cage receives a destroy event just before handling again a new HDMI output. Consequently, things are broken if terminating here.

emersion commented:
> I don't have done any tests with X11 backend for now.

It's fine, it should behave exactly just like the Wayland backend.

> When using Wayland backend, how can we destroy output? I understand with next version of wlroots, we will be able to disable outputs, but how about output destruction?

Closing windows should destroy the output, regardless of wlroots version.

> Below a short description of the issue I observed with terminate when last output available is destroyed

Yeah. Maybe it would make sense to special-case Wayland and X11.

joggee-fr commented:
> Closing windows should destroy the output, regardless of wlroots version.

As I understand it (and according to tests), in such a case, whatever the selected backend, Cage receives `sigchld_handler()` as the started/forked child process has been closed and terminates. Last available output patch does not modify this behavior.

emersion commented:
No, I'm talking about closing the wlroots Wayland/X11 backend window. The nested child process is not involved here.

I don't think Cage ever sends a `xdg_toplevel.close` event to the nested Wayland client.

emersion commented:
Come to think of it, maybe it would be better to not force-terminate when the wlroots Wayland/X11 backend window is closed by the user. Instead we could just relay the close request to the nested Wayland client. That way "do you want to save your work before closing?" kind of dialogs would work properly. But this requires some wlroots changes: https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3589

joggee-fr commented:
> No, I'm talking about closing the wlroots Wayland/X11 backend window. The nested child process is not involved here.

Ok, now I get it. I have just given it a try with Wayland backend and indeed, Cage is now running without any output available. Kind of weird 😅 .

emersion commented:
If you want to get this PR merged, maybe we can leave out that commit for now and fix it in a separate PR?

joggee-fr commented:
> If you want to get this PR merged, maybe we can leave out that commit for now and fix it in a separate PR?

Just dropped the commit from this PR.
Who get rights to merge it (if OK)? @Hjdskes does not seem to be (publicly) active here since several months.

emersion commented:
Thanks!
